### DISCUSSION TITLE
Can Jotai be used as a replacement for React Query?

### DISCUSSION THREAD

**Discussion by AjaxSolutions** [pmndrs/jotai#975] [Q&A] [answered]
I just started learning Jotai, and so far I'm very impressed.  I'm familiar with React Query.  I created a small replacement hook for the RQ `useQuery` hook, because I don't need all the features provided by RQ.

The common use case for the useQuery hook is to make a (1) request when the array `queryKey` changes and pass the params in they key to the fetcher function, (2) cache the response (cacheTime: Infinity), and (3) make a conditional request when `enabled: true`, and (4) return an object containing `{data, isLoading, isError}`.

I think all these use cases are can be covered by Jotai...?

**Comment by dai-shi** [comment]
I'm not super familiar with RQ to answer this question in general, but maybe @aulneau can.

Jotai provides basic async features, but it has some limitations compared to RQ.
- Atoms are distinguished by references
- It can have only a current value (cache size is 1 or even 0)
- No method for "refersh" or "re-evaluate" (needs a hack)

Having that said, some simple data fetching use cases could be covered by only jotai atoms.
If you have some specific patterns you want to solve, feel free to ask them.

**Reply by AjaxSolutions** [reply]
Thanks @dai-shi ,  I'll leave this question open for a while.  I'm curious if someone used ``Jotai`` to replace the``useQuery`` hook from React Query.

**Reply by dai-shi** [reply]
Just playing a little bit (not for production):
```js
const queryFamily = atomFamily(
  ({ queryFn, enabled }) => atom(() => enabled && queryFn()),
  (a, b) => a.queryKey === b.queryKey
)

const useQuery = (params) => useAtomValue(loadable(queryFamily(params)))
```

**Comment by aulneau** [comment, marked as answer]
Hi!

I think that jotai and react-query can work really, really well together. I maintain the integration in jotai, and I have this other library too: https://github.com/fungible-systems/jotai-query-toolkit, that basically gives you all the features of react-query, but in the forms of atoms. 

> The common use case for the useQuery hook is to make a (1) request when the array queryKey changes and pass the params in they key to the fetcher function, (2) cache the response (cacheTime: Infinity), and (3) make a conditional request when enabled: true, and (4) return an object containing {data, isLoading, isError}.
>
> I think all these use cases are can be covered by Jotai...?

If you want to build out some functionality yourself and not include the react-query library, you def can. The only areas that become hard are when you pass functions as params to an atom, because it's not a stable value, there can be infinite loops (cc @dai-shi , anything we can do to fix this?)

Happy to answer any questions you might have, but the example @dai-shi showed is a good first step.
